inc/vcf/record_cache.hpp

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright 2016 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef VCF_VALIDATOR_RECORDCACHE_HPP
+#define VCF_VALIDATOR_RECORDCACHE_HPP
+
+
+#include <set>
+#include <sstream>
+#include "normalizer.hpp"
+#include "file_structure.hpp"
+
+namespace ebi
+{
+  namespace vcf
+  {
+
+    /**
+     * Stores a summary of a Record to check that there are no duplicates.
+     *
+     * For memory reasons, this class can be configured to store only the last `n` elements. This will only detect
+     * duplicates if the input is almost sorted (i.e. if no element is unsorted out of its place more than `n` elements)
+     */
+    class RecordCache
+    {
+      public:
+        RecordCache() : RecordCache{1000} { }
+
+        /**
+         * @param capacity: maximum amount of RecordCores that this instance can hold at any time.
+         * Default is 1000.
+         * A value of 0 is valid but is useless.
```

Comment:
If a zero value is useless as it is, maybe it could be moved into the other branch (zero or less disables the limit).